deepinterpolation/generator_collection.py

```diff
@@ -897,112 +919,147 @@ class OphysGenerator(SequentialGenerator):
     def __init__(self, json_path):
         "Initialization"
         super().__init__(json_path)
+        self._gpu_available = tf.test.is_gpu_available()
+        self._movie_data = None
+        if self._gpu_available:
+            self._batch_tensor_index = None # Current index cached as a batch tensor on the GPU
+            self._batch_tensor = None
+            self._batch_indices = None
+            self._input_indices = None
 
         # For backward compatibility
-        if "train_path" in self.json_data.keys():
-            self.raw_data_file = self.json_data["train_path"]
+        self.cache_data = self.json_data.get("cache_data", False)
+        self.raw_data_file = self.json_data.get(
+            "train_path", self.json_data.get("movie_path"))
+        if self._gpu_available:
+            self.gpu_cache_full = self.json_data.get("gpu_cache_full", False)
         else:
-            self.raw_data_file = self.json_data["movie_path"]
-
+            self.gpu_cache_full = False
+        self.normalize_cache = self.json_data.get("normalize_cache", False)
         self.batch_size = self.json_data["batch_size"]
-        movie_obj_point = h5py.File(self.raw_data_file, "r")
-        raw_data = movie_obj_point["data"]
+        self.movie_statistics_sample_size = \
+            self.json_data.get("movie_statistics_sample_size", 1000)
 
-        self.total_frame_per_movie = int(raw_data.shape[0])
+        self.total_frame_per_movie = int(self.movie_data.shape[0])
 
         self._update_end_frame(self.total_frame_per_movie)
         self._calculate_list_samples(self.total_frame_per_movie)
 
-        average_nb_samples = np.min([int(raw_data.shape[0]), 100])
 
-        # For backward compatibility
-        if "cache_data" in self.json_data.keys():
-            self.cache_data = self.json_data["cache_data"]
-        else:
-            self.cache_data = False
-
-        if self.cache_data:
-            print('Caching hdf5 file... \n')
-            if self.end_frame > 0:
-                self.raw_data = raw_data[0:np.min([self.total_frame_per_movie,
-                    self.end_frame+self.post_frame+self.pre_post_omission+1]), :, :]
-            else:
-                self.raw_data = raw_data[:, :, :]
-            local_data = self.raw_data[0:average_nb_samples, :, :].flatten()
-        else:
-            local_data = raw_data[0:average_nb_samples, :, :].flatten()
-            self.cache_data = False
-
-        local_data = local_data.astype("float32")
-
-        self.local_mean = np.mean(local_data)
-        self.local_std = np.std(local_data)
+    @property
+    def movie_data(self) -> Union[np.ndarray, tf.Tensor]:
+        if self._movie_data is None:
+            with h5py.File(self.raw_data_file, "r") as movie_obj:
```

Comment:
As far as I understand, this code now forces everyone to load the entire set of frames in RAM no matter what. The generator was written to allow lazy loading so that people with minimal RAM could do this OR we could handle very very large movies. I would prefer to allow lazy loading. The previous generator was also written to repoint at a file when fetching batches.

Reply:
Given that the entire generator was rewritten, I have to check its behavior as this code was the result of months of optimization and edge-cases exploration. Isn't there a simpler way to add the GPU loading?